constraint-solver/src/solver/exhaustive_search.rs

```diff
@@ -85,23 +85,28 @@ fn find_unique_assignment_for_set<T: FieldElement, V: Clone + Hash + Ord + Eq +
     rc: impl RangeConstraintProvider<T, V> + Clone,
     bus_interaction_handler: &impl BusInteractionHandler<T>,
 ) -> Result<Option<BTreeMap<V, T>>, Error> {
-    match get_all_possible_assignments(variables.iter().cloned(), &rc)
-        .filter(|assignments| {
-            !constraint_system.is_assignment_conflicting(assignments, &rc, bus_interaction_handler)
-        })
-        .exactly_one()
-    {
-        Ok(assignments) => Ok(Some(assignments)),
-        Err(mut iter) => {
-            if iter.next().is_some() {
-                // There are at least two assignments, so there is no unique assignment.
-                Ok(None)
-            } else {
-                // No assignment satisfied the constraint system.
-                Err(Error::ExhaustiveSearchError)
-            }
+    let mut assignments =
+        get_all_possible_assignments(variables.iter().cloned(), &rc).filter_map(|assignments| {
+            constraint_system
+                .derive_more_assignments(assignments, &rc, bus_interaction_handler)
+                .ok()
+        });
+    let Some(first_assignments) = assignments.next() else {
```

Comment:
What about using `assignments.reduce(...)` and then return an error if it's none?

Reply:
Hmm, nice idea, but `reduce()` does not let me exit early, which is crucial for performance!